File: Code/MaxAlHelper/Triggers/OmniDirectionalSnowballTrigger.cs
namespace MaxAlHelper;


[CustomEntity("MaxAlHelper/OmniDirectionalSnowballTrigger")]
public class OmniDirectionalSnowballTrigger : Trigger {
    public float Speed;
    public float ResetTime;
    public bool DrawOutline;
    public string SpritePath;
    public float SineWaveFrequency;
    public float AppearAngle;
    public bool ReplaceExisting;
    public float SafeZoneSize;
    public float Offset;

    public OmniDirectionalSnowballTrigger(EntityData data, Vector2 offset) : base(data, offset) {
        SpritePath = data.Attr("spritePath", "snowball");
        Speed = data.Float("speed", 200f);
        ResetTime = data.Float("resetTime", 0.8f);
        SineWaveFrequency = data.Float("ySineWaveFrequency", 0.5f);
        DrawOutline = data.Bool("drawOutline", true);
        AppearAngle = data.Float("appearAngle", 0f);
        ReplaceExisting = data.Bool("replaceExisting", true);
        SafeZoneSize = data.Float("safeZoneSize", 64f);
        Offset = data.Float("offset", 0f);
    }

    public override void OnEnter(Player player) {
        base.OnEnter(player);
        OmniDirectionalSnowball snowball;
        if (ReplaceExisting && (snowball = Scene.Entities.FindFirst<OmniDirectionalSnowball>()) != null) {
            snowball.Speed = Speed;
            snowball.ResetTime = ResetTime;
            snowball.Sine.Frequency = SineWaveFrequency;
            if (snowball.Sprite.Path != SpritePath) {
                snowball.CreateSprite(SpritePath);
            }
            snowball.DrawOutline = DrawOutline;
            snowball.appearAngle = AppearAngle;
            snowball.SafeZoneSize = SafeZoneSize;
        } else {
            Scene.Add(new OmniDirectionalSnowball(SpritePath, Speed, ResetTime, SineWaveFrequency, DrawOutline, AppearAngle, SafeZoneSize, Offset));
        }
        RemoveSelf();
    }
}

File: Loenn/triggers/OmniDirectionalSnowballTrigger.lua
local omniDirectionalSnowballTrigger = {}

-- remember to match the name with the CustomEntity plugin name
omniDirectionalSnowballTrigger.name = "MaxAlHelper/OmniDirectionalSnowballTrigger"

-- you don't need an extra layer of braces if you're defining only one placement
-- this is fine though, no need to edit
omniDirectionalSnowballTrigger.placements = {
    {
        name = "normal",
        data = {
            spritePath = "snowball",
            appearAngle = 0.0,
            speed = 200.0,
            resetTime = 0.8,
            ySineWaveFrequency = 0.5,
            safeZoneSize = 64.0,
            offset = 0.0,
            drawOutline = true,
            replaceExisting = true,
            repetitionsForTurn = 0,
            turnAngle = 0.0
        }
    }
}

return omniDirectionalSnowballTrigger
